File: golang/pkg/datastructures/array.go
package datastructures

import (
	"encoding/binary"
	"io"
)

// Array - динамический массив строк
type Array struct {
	data     []string
	size     int
	capacity int
}

// NewArray создаёт новый пустой массив
func NewArray() *Array {
	return &Array{
		data:     make([]string, 0),
		size:     0,
		capacity: 0,
	}
}

// IsEmpty проверяет, пуст ли массив
func (a *Array) IsEmpty() bool {
	return a.size == 0
}

// Reserve выделяет память для массива
func (a *Array) Reserve(newCapacity int) {
	if newCapacity <= a.capacity {
		return
	}
	newData := make([]string, newCapacity)
	copy(newData, a.data[:a.size])
	a.data = newData
	a.capacity = newCapacity
}

// PushBack добавляет элемент в конец
func (a *Array) PushBack(val string) {
	if a.size == a.capacity {
		if a.capacity == 0 {
			a.Reserve(4)
		} else {
			a.Reserve(a.capacity * 2)
		}
	}
	a.data[a.size] = val
	a.size++
}

// InsertAt вставляет элемент по индексу
func (a *Array) InsertAt(index int, val string) error {
	if index < 0 || index > a.size {
		return ErrIndexOutOfRange
	}

	if a.size == a.capacity {
		if a.capacity == 0 {
			a.Reserve(4)
		} else {
			a.Reserve(a.capacity * 2)
		}
	}

	for i := a.size; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = val
	a.size++
	return nil
}

// DelAt удаляет элемент по индексу
func (a *Array) DelAt(index int) error {
	if index < 0 || index >= a.size {
		return ErrIndexOutOfRange
	}

	for i := index; i < a.size-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.size--
	return nil
}

// GetAt получает элемент по индексу
func (a *Array) GetAt(index int) (string, error) {
	if index < 0 || index >= a.size {
		return "", ErrIndexOutOfRange
	}
	return a.data[index], nil
}

// ReplaceAt заменяет элемент по индексу
func (a *Array) ReplaceAt(index int, val string) error {
	if index < 0 || index >= a.size {
		return ErrIndexOutOfRange
	}
	a.data[index] = val
	return nil
}

// GetSize возвращает размер массива
func (a *Array) GetSize() int {
	return a.size
}

// Serialize сохраняет массив в бинарном формате
func (a *Array) Serialize(w io.Writer) error {
	// Сохраняем размер
	if err := binary.Write(w, binary.LittleEndian, int32(a.size)); err != nil {
		return err
	}

	// Сохраняем каждый элемент
	for i := 0; i < a.size; i++ {
		len := int32(len(a.data[i]))
		if err := binary.Write(w, binary.LittleEndian, len); err != nil {
			return err
		}
		if _, err := w.Write([]byte(a.data[i])); err != nil {
			return err
		}
	}
	return nil
}

// Deserialize загружает массив из бинарного формата
func (a *Array) Deserialize(r io.Reader) error {
	// Очищаем текущие данные
	a.data = make([]string, 0)
	a.size = 0
	a.capacity = 0

	// Загружаем размер
	var size int32
	if err := binary.Read(r, binary.LittleEndian, &size); err != nil {
		return err
	}

	// Загружаем элементы
	for i := 0; i < int(size); i++ {
		var len int32
		if err := binary.Read(r, binary.LittleEndian, &len); err != nil {
			return err
		}

		buf := make([]byte, len)
		if _, err := r.Read(buf); err != nil {
			return err
		}

		a.PushBack(string(buf))
	}
	return nil
}

File: sd/array/array.cpp
#include "array.hpp"
#include <iostream>
using namespace std;

// Конструктор
Mass::Mass() : data(nullptr), size(0), capacity(0) {}

// Деструктор
Mass::~Mass() { delete[] data; }

// Проверка на пустоту
bool Mass::is_empty() const { return size == 0; }

// Предварительное выделение памяти
void Mass::reserve(int new_capacity) {
  if (new_capacity <= capacity)
    return;
  string *new_data = new string[new_capacity];
  for (int i = 0; i < size; ++i) {
    new_data[i] = data[i];
  }
  delete[] data;
  data = new_data;
  capacity = new_capacity;
}

// Добавление элемента в конец
void Mass::push_back(const string &val) {
  if (size == capacity) {
    reserve(capacity == 0 ? 4 : capacity * 2);
  }
  data[size++] = val;
}

// Вставка элемента по индексу
void Mass::insert_at(int index, const string &val) {
  if (index < 0 || index > size)
    return;
  if (size == capacity) {
    reserve(capacity == 0 ? 4 : capacity * 2);
  }
  for (int i = size; i > index; --i) {
    data[i] = data[i - 1];
  }
  data[index] = val;
  ++size;
}

// Удаление элемента по индексу
void Mass::del_at(int index) {
  if (index < 0 || index >= size)
    return;
  for (int i = index; i < size - 1; ++i) {
    data[i] = data[i + 1];
  }
  --size;
}

// Получение элемента по индексу
string Mass::get_at(int index) const {
  if (index < 0 || index >= size)
    return "";
  return data[index];
}

// Замена элемента по индексу
void Mass::replace_at(int index, const string &val) {
  if (index < 0 || index >= size)
    return;
  data[index] = val;
}

// Получение размера массива
int Mass::get_size() const { return size; }

// Печать всех элементов
void Mass::print() const {
  for (int i = 0; i < size; ++i) {
    cout << data[i] << " ";
  }
  cout << endl;
}

// Считывание элементов с консоли
void Mass::read() {
  int n;
  cout << "Введите количество элементов: ";
  cin >> n;
  cin.ignore();

  for (int i = 0; i < n; ++i) {
    string val;
    cout << "Элемент " << i << ": ";
    getline(cin, val);
    push_back(val);
  }
}

// Бинарная сериализация
void Mass::serialize(std::ostream &out) const {
  out.write(reinterpret_cast<const char *>(&size), sizeof(int));
  for (int i = 0; i < size; ++i) {
    int len = data[i].length();
    out.write(reinterpret_cast<const char *>(&len), sizeof(int));
    out.write(data[i].c_str(), len);
  }
}

// Бинарная десериализация
void Mass::deserialize(std::istream &in) {
  int new_size = 0;
  in.read(reinterpret_cast<char *>(&new_size), sizeof(int));

  // Очищаем текущие данные
  delete[] data;
  data = nullptr;
  size = 0;
  capacity = 0;

  // Загружаем элементы
  for (int i = 0; i < new_size; ++i) {
    int len = 0;
    in.read(reinterpret_cast<char *>(&len), sizeof(int));

    char *buffer = new char[len + 1];
    in.read(buffer, len);
    buffer[len] = '\0';

    string val(buffer);
    delete[] buffer;
    push_back(val);
  }
}